testcase/unittest/blockchain/transactions/test_transaction_builder.py

```diff
@@ -0,0 +1,181 @@
+        assert not tx_builder.signature
+        assert not tx_builder.hash
+
+        tx_builder.build_from_address()
+        tx_builder.build_origin_data()
+
+        mocker.patch.object(Signer, "sign_hash", return_value=os.urandom(Signature.size))
+        tx_builder.sign()
+        assert isinstance(tx_builder.signature, Signature)
+        assert tx_builder.hash
+
+    def test_build_raw_data(self, tx_builder_factory, tx_version):
+        tx_builder = tx_builder_factory(tx_version=tx_version)
+
+        tx_builder.build_from_address()
+        tx_builder.build_origin_data()
+        tx_builder.build_hash()
+        tx_builder.sign()
+        raw_data = tx_builder.build_raw_data()
+
+        assert raw_data
+
+    def test_build_transaction(self, tx_builder_factory, tx_version):
+        tx_builder = tx_builder_factory(tx_version=tx_version)
+        tx = tx_builder.build()
+
+        if tx_version == genesis.version:
+            assert isinstance(tx, genesis.Transaction)
+        elif tx_version == v2.version:
+            assert isinstance(tx, v2.Transaction)
+        elif tx_version == v3.version:
+            assert isinstance(tx, v3.Transaction)
+
+    def test_sign_transaction(self, tx_builder_factory, tx_version):
+        tx_builder = tx_builder_factory(tx_version=tx_version)
+        tx = tx_builder.build()
+
+        if tx_version == genesis.version:
+            with pytest.raises(NotImplementedError):
+                tx_builder.sign_transaction(tx)
+        else:
+            tx_builder.sign_transaction(tx)
+
+    def test_sign_transaction_raises_exc_if_signer_addr_ne_tx_signer_addr(self, tx_builder_factory, tx_version):
+        tx_builder = tx_builder_factory(tx_version=tx_version)
+        tx = tx_builder.build()
+
+        tx_builder.signer = Signer.new()
+
+        if tx_version == genesis.version:
+            with pytest.raises(NotImplementedError):
+                tx_builder.sign_transaction(tx)
+        else:
+            with pytest.raises(RuntimeError, match="Signer not match"):
+                tx_builder.sign_transaction(tx)
+
+
+class TestTransactionBuilder_genesis:
+    def test_build_nid_raises_exc_if_hash_not_exists(self, tx_builder_factory):
+        tx_builder: genesis.TransactionBuilder = tx_builder_factory(tx_version=genesis.version)
+        tx_builder.reset_cache()
+
+        with pytest.raises(RuntimeError):
+            assert tx_builder.build_nid()
+
+    def test_build_nid_returns_its_nid_if_exists(self, tx_builder_factory):
+        tx_builder: genesis.TransactionBuilder = tx_builder_factory(tx_version=genesis.version)
+        expected_nid = genesis.NID.unknown
+        tx_builder.nid = expected_nid
+        tx_builder.build_origin_data()
+        tx_builder.build_hash()
+        nid_generated = tx_builder.build_nid()
+
+        assert expected_nid == nid_generated
+
+    @pytest.mark.parametrize("tx_hash, expected_nid", [
+        (genesis.NTxHash.mainnet, genesis.NID.mainnet),
```

Comment:
These values are written here inline.
Because the tests will pass even if these values are changed from original codes.